include/up-cpp/utils/Expected.h

```diff
@@ -27,82 +28,137 @@ static_assert(!__has_cpp_attribute(__cpp_lib_expected),
 ///          https://en.cppreference.com/w/cpp/utility/expected
 ///          No further documentation is provided in this file.
 /// @{
-template <typename E>
-struct BadExpectedAccess;
+struct BadExpectedAccess : public std::runtime_error {
+	BadExpectedAccess() : std::runtime_error("generic BadExpectedAccess") {}
 
-template <>
-struct BadExpectedAccess<void> : public std::exception {};
+	template <typename... Args>
+	BadExpectedAccess(Args&&... args)
+	    : std::runtime_error(std::forward<Args>(args)...) {}
+};
 
+/// @brief Required tagging type for cases where expected and unexpected type
+/// are identical.
 template <typename E>
-struct BadExpectedAccess : public BadExpectedAccess<void> {};
+class Unexpected {
+public:
+	constexpr Unexpected(const Unexpected&) = default;
+	constexpr Unexpected(Unexpected&&) = default;
+
+	constexpr explicit Unexpected(const E& rhs) : storage(rhs) {}
+	constexpr explicit Unexpected(E&& rhs) : storage(std::move(rhs)) {}
+
+	template <class... Args>
+	constexpr explicit Unexpected(std::in_place_t, Args&&... args)
+	    : storage(std::forward<Args>(args)...) {}
+
+	constexpr const E& error() const& noexcept { return storage; }
+
+	constexpr const E&& error() const&& noexcept { return storage; }
+
+	constexpr E&& error() && noexcept { return std::move(storage); }
+
+private:
+	E storage;
+};
 
 /// @brief A stripped-down version of std::expected from C++23.
 template <typename T, typename E>
-struct Expected {
-	using value_type = T;
-	using error_type = E;
-
-	constexpr Expected();
-	constexpr Expected(const Expected& other);
-	constexpr Expected(Expected&& other) noexcept;
-
-	constexpr explicit Expected(T&& v);
-	constexpr explicit Expected(E&& e);
-
-	Expected& operator=(const Expected& other);
-	Expected& operator=(Expected&& other);
-
-	Expected& operator=(const T& other);
-	Expected& operator=(T&& other);
-
-	Expected& operator=(const E& other);
-	Expected& operator=(E&& other);
-
-	constexpr explicit operator bool() const noexcept;
-	constexpr bool has_value() const noexcept;
-
-	constexpr T& value() &;
-	constexpr const T& value() const&;
-	constexpr T&& value() &&;
-	constexpr const T&& value() const&&;
-
-	constexpr const E& error() const& noexcept;
-	constexpr E& error() & noexcept;
-	constexpr const E&& error() const&& noexcept;
-	constexpr E&& error() && noexcept;
-
-	constexpr T value_or(T&& default_value) const&;
-	constexpr T value_or(T&& default_value) &&;
-
-	template <typename F>
-	constexpr auto and_then(F&& f) &;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&;
-	template <typename F>
-	constexpr auto and_then(F&& f) &&;
-	template <typename F>
-	constexpr auto and_then(F&& f) const&&;
-
-	template <typename F>
-	constexpr auto or_else(F&& f) &;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&;
-	template <typename F>
-	constexpr auto or_else(F&& f) &&;
-	template <typename F>
-	constexpr auto or_else(F&& f) const&&;
+class Expected {
+public:
```

Comment:
@debruce - You may disagree with this, but for consistency with the current codebase _please_ make this change. This is not the first time I have made this _exact_ request in this PR.

Once a coding standard has been reviewed and adopted, this may change.

Reply:
There are things I care more about in this review that have not been addressed. This can be evaluated when the coding standard is adopted.